File: scripts/eval-voice-character-suite.mjs
import fs from "fs";
import path from "path";
import { parseText, synthesizeHumanizedToMp3 } from "../lib/tts-core.mjs";

function parseArgs(argv) {
  const out = {};
  for (let i = 0; i < argv.length; i += 1) {
    const item = argv[i];
    if (!item.startsWith("--")) continue;
    const key = item.slice(2);
    const next = argv[i + 1];
    if (!next || next.startsWith("--")) {
      out[key] = true;
      continue;
    }
    out[key] = next;
    i += 1;
  }
  return out;
}

function average(items) {
  if (!items.length) return 0;
  return items.reduce((s, n) => s + n, 0) / items.length;
}

function readTestText(dirPath) {
  const files = fs
    .readdirSync(dirPath, { withFileTypes: true })
    .filter((d) => d.isFile() && d.name.toLowerCase().endsWith(".txt"))
    .map((d) => d.name)
    .sort();
  if (!files.length) {
    throw new Error(`No .txt voice test files found in ${dirPath}`);
  }
  const first = files[0];
  const raw = fs.readFileSync(path.join(dirPath, first), "utf8");
  const parsed = parseText(raw);
  return {
    file: first,
    text: parsed.text || raw
  };
}

async function runCase({
  text,
  outputBase,
  voice,
  style,
  speechStyle,
  voiceTone
}) {
  const res = await synthesizeHumanizedToMp3({
    text,
    output: outputBase,
    voice,
    rate: "-8%",
    pitch: "-2Hz",
    volume: "0%",
    humanizeIntensity: 0.55,
    style,
    speechStyle,
    useMlPolicy: false,
    autoExpressive: true,
    voiceCharacter: true,
    voiceTone
  });
  const prosody = JSON.parse(fs.readFileSync(res.prosodyPath, "utf8"));
  const seg = Array.isArray(prosody.segments) ? prosody.segments : [];
  return {
    voice,
    style: prosody.style,
    speechStyle,
    voiceTone,
    outputAudio: path.relative(process.cwd(), res.audioPath),
    outputProsody: path.relative(process.cwd(), res.prosodyPath),
    segments: seg.length,
    avgRate: Number(average(seg.map((s) => Number(s?.final?.rate ?? s?.rate ?? 0))).toFixed(3)),
    avgPitch: Number(average(seg.map((s) => Number(s?.final?.pitch ?? s?.pitch ?? 0))).toFixed(3)),
    postApplied: Boolean(prosody?.voiceCharacter?.postApplied)
  };
}

function boolAll(items) {
  return items.every(Boolean);
}

async function main() {
  const args = parseArgs(process.argv.slice(2));
  const testsDir = path.resolve(String(args.dir || "tests/voice"));
  const outDir = path.resolve(String(args.outdir || "outputs/voice_eval_suite"));
  const style = String(args.style || "tegang");
  const minRateDelta = Number(args["min-rate-delta"] ?? 1.0);
  const voices = String(args.voices || "id-ID-ArdiNeural,id-ID-GadisNeural")
    .split(",")
    .map((v) => String(v || "").trim())
    .filter(Boolean);

  if (!fs.existsSync(testsDir)) {
    throw new Error(`Voice tests directory not found: ${testsDir}`);
  }
  if (!fs.existsSync(outDir)) fs.mkdirSync(outDir, { recursive: true });

  const sample = readTestText(testsDir);
  const configs = [
    { speechStyle: "relaxed", voiceTone: "auto" },
    { speechStyle: "assertive", voiceTone: "auto" },
    { speechStyle: "dramatic", voiceTone: "deep" },
    { speechStyle: "dramatic", voiceTone: "bright" },
    { speechStyle: "dramatic", voiceTone: "off" }
  ];

  const rows = [];
  for (const voice of voices) {
    for (const cfg of configs) {
      const base = path.join(
        outDir,
        `${voice.replace(/[^a-z0-9]/gi, "_")}_${cfg.speechStyle}_${cfg.voiceTone}`
      );
      const result = await runCase({
        text: sample.text,
        outputBase: base,
        voice,
        style,
        speechStyle: cfg.speechStyle,
        voiceTone: cfg.voiceTone
      });
      rows.push(result);
      console.log(
        `voice_case voice=${voice} speech_style=${cfg.speechStyle} voice_tone=${cfg.voiceTone} post=${result.postApplied} avg_rate=${result.avgRate}`
      );
    }
  }

  const checks = [];
  for (const voice of voices) {
    const relaxed = rows.find((r) => r.voice === voice && r.speechStyle === "relaxed" && r.voiceTone === "auto");
    const assertive = rows.find(
      (r) => r.voice === voice && r.speechStyle === "assertive" && r.voiceTone === "auto"
    );
    const deep = rows.find((r) => r.voice === voice && r.voiceTone === "deep");
    const bright = rows.find((r) => r.voice === voice && r.voiceTone === "bright");
    const off = rows.find((r) => r.voice === voice && r.voiceTone === "off");

    const rateDelta = Number(((assertive?.avgRate ?? 0) - (relaxed?.avgRate ?? 0)).toFixed(3));
    const perVoice = {
      voice,
      relaxedVsAssertiveRateDelta: rateDelta,
      relaxedVsAssertiveOk: rateDelta >= minRateDelta,
      deepPostOk: Boolean(deep?.postApplied),
      brightPostOk: Boolean(bright?.postApplied),
      offPostDisabledOk: !Boolean(off?.postApplied)
    };
    checks.push(perVoice);
  }

  const summary = {
    at: new Date().toISOString(),
    testsDir: path.relative(process.cwd(), testsDir),
    sampleFile: sample.file,
    style,
    minRateDelta,
    caseCount: rows.length,
    voices,
    checks,
    gate: {
      status: boolAll(
        checks.map(
          (c) => c.relaxedVsAssertiveOk && c.deepPostOk && c.brightPostOk && c.offPostDisabledOk
        )
      )
        ? "pass"
        : "fail"
    },
    cases: rows
  };

  const jsonPath = path.join(outDir, "summary.json");
  fs.writeFileSync(jsonPath, JSON.stringify(summary, null, 2), "utf8");

  const md = [
    "# Voice Character Eval Suite",
    "",
    `- Generated at: ${summary.at}`,
    `- Test dir: \`${summary.testsDir}\``,
    `- Sample file: \`${summary.sampleFile}\``,
    `- Style: \`${summary.style}\``,
    `- Gate: **${summary.gate.status.toUpperCase()}**`,
    "",
    "## Voice Checks",
    "",
    "| voice | rateDelta(assertive-relaxed) | rateOk | deepPost | brightPost | offDisabled |",
    "|---|---:|---|---|---|---|"
  ];
  for (const c of checks) {
    md.push(
      `| ${c.voice} | ${c.relaxedVsAssertiveRateDelta} | ${c.relaxedVsAssertiveOk} | ${c.deepPostOk} | ${c.brightPostOk} | ${c.offPostDisabledOk} |`
    );
  }
  const mdPath = path.join(outDir, "summary.md");
  fs.writeFileSync(mdPath, `${md.join("\n")}\n`, "utf8");

  console.log(
    `voice_eval_done gate=${summary.gate.status} cases=${rows.length} summary_json=${path.relative(process.cwd(), jsonPath)} summary_md=${path.relative(process.cwd(), mdPath)}`
  );
  if (summary.gate.status !== "pass") process.exitCode = 2;
}

main().catch((err) => {
  console.error(err.message || String(err));
  process.exit(1);
});

File: scripts_py/eval_voice_character_suite.py
import json
import os
import tempfile
from datetime import datetime, timezone
from pathlib import Path

from common_pipeline import parse_args, run_cmd


def average(items):
    return (sum(items) / len(items)) if items else 0.0


def read_test_text(dir_path):
    files = sorted([p for p in Path(dir_path).iterdir() if p.is_file() and p.suffix.lower() == ".txt"])
    if not files:
        raise RuntimeError(f"No .txt voice test files found in {dir_path}")
    fp = files[0]
    raw = fp.read_text(encoding="utf-8")
    lines = [ln.strip() for ln in raw.splitlines() if ln.strip() and not ln.strip().startswith("[SCENE")]
    text = "\n".join(lines).strip() or raw
    return {"file": fp.name, "text": text}


def run_case(text, output_base, voice, style, speech_style, voice_tone):
    with tempfile.NamedTemporaryFile(prefix="voice_case_", suffix=".txt", delete=False) as tmp:
        tmp.write((text + "\n").encode("utf-8"))
        tmp_path = tmp.name
    try:
        cmd = [
            "node",
            "scripts/generate-tts.mjs",
            "--input",
            tmp_path,
            "--output",
            output_base,
            "--voice",
            voice,
            "--humanize",
            "true",
            "--humanize-intensity",
            "0.55",
            "--style",
            style,
            "--speech-style",
            speech_style,
            "--voice-character",
            "true",
            "--voice-tone",
            voice_tone,
            "--ml-policy",
            "false",
        ]
        run_cmd(cmd, allowed_exit_codes=(0,))
        prosody_path = Path(f"{output_base}.prosody.json").resolve()
        audio_path = Path(f"{output_base}.mp3").resolve()
        with open(prosody_path, "r", encoding="utf-8") as f:
            prosody = json.load(f)
        seg = prosody.get("segments") if isinstance(prosody, dict) else []
        seg = seg if isinstance(seg, list) else []
        rates = [float((s.get("final") or {}).get("rate", s.get("rate", 0))) for s in seg]
        pitches = [float((s.get("final") or {}).get("pitch", s.get("pitch", 0))) for s in seg]
        return {
            "voice": voice,
            "style": prosody.get("style"),
            "speechStyle": speech_style,
            "voiceTone": voice_tone,
            "outputAudio": str(audio_path.relative_to(Path.cwd())).replace("\\", "/"),
            "outputProsody": str(prosody_path.relative_to(Path.cwd())).replace("\\", "/"),
            "segments": len(seg),
            "avgRate": round(average(rates), 3),
            "avgPitch": round(average(pitches), 3),
            "postApplied": bool(((prosody.get("voiceCharacter") or {}).get("postApplied"))),
        }
    finally:
        try:
            os.remove(tmp_path)
        except Exception:
            pass


def bool_all(items):
    return all(bool(x) for x in items)


def main():
    args = parse_args(os.sys.argv[1:])
    tests_dir = Path(str(args.get("dir") or "tests/voice")).resolve()
    out_dir = Path(str(args.get("outdir") or "outputs/voice_eval_suite")).resolve()
    style = str(args.get("style") or "tegang")
    min_rate_delta = float(args.get("min-rate-delta") or 1.0)
    voices = [v.strip() for v in str(args.get("voices") or "id-ID-ArdiNeural,id-ID-GadisNeural").split(",") if v.strip()]
    if not tests_dir.exists():
        raise RuntimeError(f"Voice tests directory not found: {tests_dir}")
    out_dir.mkdir(parents=True, exist_ok=True)

    sample = read_test_text(tests_dir)
    configs = [
        {"speechStyle": "relaxed", "voiceTone": "auto"},
        {"speechStyle": "assertive", "voiceTone": "auto"},
        {"speechStyle": "dramatic", "voiceTone": "deep"},
        {"speechStyle": "dramatic", "voiceTone": "bright"},
        {"speechStyle": "dramatic", "voiceTone": "off"},
    ]

    rows = []
    for voice in voices:
        for cfg in configs:
            base = out_dir / f"{''.join(c if c.isalnum() else '_' for c in voice)}_{cfg['speechStyle']}_{cfg['voiceTone']}"
            result = run_case(
                text=sample["text"],
                output_base=str(base),
                voice=voice,
                style=style,
                speech_style=cfg["speechStyle"],
                voice_tone=cfg["voiceTone"],
            )
            rows.append(result)
            print(
                f"voice_case voice={voice} speech_style={cfg['speechStyle']} voice_tone={cfg['voiceTone']} "
                f"post={str(result['postApplied']).lower()} avg_rate={result['avgRate']}"
            )

    checks = []
    for voice in voices:
        relaxed = next((r for r in rows if r["voice"] == voice and r["speechStyle"] == "relaxed" and r["voiceTone"] == "auto"), None)
        assertive = next((r for r in rows if r["voice"] == voice and r["speechStyle"] == "assertive" and r["voiceTone"] == "auto"), None)
        deep = next((r for r in rows if r["voice"] == voice and r["voiceTone"] == "deep"), None)
        bright = next((r for r in rows if r["voice"] == voice and r["voiceTone"] == "bright"), None)
        off = next((r for r in rows if r["voice"] == voice and r["voiceTone"] == "off"), None)
        rate_delta = round(
            float((assertive or {}).get("avgRate", 0)) - float((relaxed or {}).get("avgRate", 0)),
            3,
        )
        checks.append(
            {
                "voice": voice,
                "relaxedVsAssertiveRateDelta": rate_delta,
                "relaxedVsAssertiveOk": rate_delta >= min_rate_delta,
                "deepPostOk": bool((deep or {}).get("postApplied")),
                "brightPostOk": bool((bright or {}).get("postApplied")),
                "offPostDisabledOk": not bool((off or {}).get("postApplied")),
            }
        )

    summary = {
        "at": datetime.now(timezone.utc).isoformat(),
        "testsDir": str(tests_dir.relative_to(Path.cwd())).replace("\\", "/"),
        "sampleFile": sample["file"],
        "style": style,
        "minRateDelta": min_rate_delta,
        "caseCount": len(rows),
        "voices": voices,
        "checks": checks,
        "gate": {
            "status": "pass"
            if bool_all(
                [
                    c["relaxedVsAssertiveOk"] and c["deepPostOk"] and c["brightPostOk"] and c["offPostDisabledOk"]
                    for c in checks
                ]
            )
            else "fail"
        },
        "cases": rows,
    }
    json_path = out_dir / "summary.json"
    with open(json_path, "w", encoding="utf-8") as f:
        json.dump(summary, f, indent=2, ensure_ascii=False)
    md = [
        "# Voice Character Eval Suite",
        "",
        f"- Generated at: {summary['at']}",
        f"- Test dir: `{summary['testsDir']}`",
        f"- Sample file: `{summary['sampleFile']}`",
        f"- Style: `{summary['style']}`",
        f"- Gate: **{summary['gate']['status'].upper()}**",
        "",
        "## Voice Checks",
        "",
        "| voice | rateDelta(assertive-relaxed) | rateOk | deepPost | brightPost | offDisabled |",
        "|---|---:|---|---|---|---|",
    ]
    for c in checks:
        md.append(
            f"| {c['voice']} | {c['relaxedVsAssertiveRateDelta']} | {c['relaxedVsAssertiveOk']} | "
            f"{c['deepPostOk']} | {c['brightPostOk']} | {c['offPostDisabledOk']} |"
        )
    md_path = out_dir / "summary.md"
    with open(md_path, "w", encoding="utf-8") as f:
        f.write("\n".join(md) + "\n")
    print(
        f"voice_eval_done gate={summary['gate']['status']} cases={len(rows)} "
        f"summary_json={str(json_path.relative_to(Path.cwd())).replace('\\', '/')} "
        f"summary_md={str(md_path.relative_to(Path.cwd())).replace('\\', '/')}"
    )
    if summary["gate"]["status"] != "pass":
        raise SystemExit(2)


if __name__ == "__main__":
    main()
